### PR TITLE
Issue1249 typescript jest test framework

### LINKED ISSUE #1249: Setup TypeScript test environment
Add a way to unit test the TypeScript files.

### PR DESCRIPTION
**Pull request by BramHeijmink**
## Summary

This PR can be summarized in the following changelog entry:
Changed testing framework to Jest.

## Relevant technical choices:

* Changed testing framework to Jest.

## Test instructions

This PR can be tested by following these steps:

* Check if the existing tests still work and change a few tests to see if the tests fail.
* Create sum.ts in the /src folder and insert the code given below.
`export function sum(a: number, b: number): number {
    return a + b;
}`

And create a sumSpec.ts file in the /spec folder with the following test:
```
const sum = require('../src/sum');

it('adds 1 + 2 to equal 3 in TScript', ()=> {
    expect(sum(1, 2)).toBe(3);
});
```

Fixes #1249

### CODE REVIEW

File: tsconfig.json [hunk 1]
```diff
@@ -3,8 +3,8 @@
         "typeRoots": [
             "node_modules/@types"
         ],
-        "module": "es6",
-        "moduleResolution": "node",
+        "module": "commonjs",
+        "moduleResolution": "Node",
```
Comment (IreneStr) [inline tsconfig.json]: Sure about the capital 'n'?

File: tsconfig.json [hunk 2]
```diff
@@ -19,7 +19,8 @@
         "forceConsistentCasingInFileNames": true,
         "noEmitOnError": true,
         "noImplicitAny": true,
-        "pretty": true
+        "pretty": true,
+        "skipLibCheck": true
     },
     "filesGlob": [
         "src/**/*.ts",
```
Comment (IreneStr) [inline tsconfig.json]: not tsx?

Comment (IreneStr) [reply tsconfig.json]: (not in this pr, I know)

File: tslint.json [hunk 3]
```diff
@@ -0,0 +1,9 @@
+{
+    "defaultSeverity": "error",
+    "extends": [
+        "tslint:recommended"
+    ],
+    "jsRules": {},
+    "rules": {},
+    "rulesDirectory": []
+}
```
Comment (IreneStr) [inline tslint.json]: Please add newline on end of file.

File: tslint.json [hunk 4]
```diff
@@ -0,0 +1,9 @@
+{
+    "defaultSeverity": "error",
+    "extends": [
+        "tslint:recommended"
+    ],
+    "jsRules": {},
```
Comment (IreneStr) [inline tslint.json]: Please remove empty objects and arrays

File: tsconfig.json (hunk 1 above)
Comment (atimmer) [reply tsconfig.json]: Does this mean we can never move to ES6, or does it mean that TS compiles to commonjs modules?

File: .travis.yml [hunk 5]
```diff
@@ -23,7 +23,7 @@ before_script:
 script:
 - yarn test
 - grunt check
-- yarn run check-coverage
+- tslint 'src/**/*.ts?(x)'
```
[the remark this answers is not in the document]
Comment (BramHeijmink) [reply .travis.yml]: Enables us to use react.

File: spec/researches/getParagraphLengthSpec.js [hunk 6]
```diff
@@ -69,12 +69,12 @@ describe("a test for getting paragraph length", function(){
 		expect( getParagraphLength( mockPaper )[ 0 ].wordCount ).toBe( 5 );
 	});
 
-	it("returns the paragraph length of an empty paragraph with p tags", function(){
+	xit("returns the paragraph length of an empty paragraph with p tags", function(){
```
[the remark this answers is not in the document]
Comment (BramHeijmink) [reply spec/researches/getParagraphLengthSpec.js]: The test does not work. The return value of an empty paragraph is an empty array. It is not possible to get the word count out of an empty paragraph.

File: spec/researches/getParagraphLengthSpec.js [hunk 7]
```diff
@@ -69,12 +69,12 @@ describe("a test for getting paragraph length", function(){
...
 		var mockPaper = new Paper( "<p></p>" );
 		expect( getParagraphLength( mockPaper ).wordCount ).not.toContain( 0 );
 	});
 
-	it("returns the paragraph length of an empty paragraph without p tags or double line breaks", function(){
+	xit("returns the paragraph length of an empty paragraph without p tags or double line breaks", function(){
```
[the remark this answers is not in the document]
Comment (BramHeijmink) [reply spec/researches/getParagraphLengthSpec.js]: The test does not work. The return value of an empty paragraph is an empty array. It is not possible to get the word count out of an empty paragraph.

File: package.json [hunk 8]
```diff
@@ -16,14 +16,17 @@
   },
   "scripts": {
     "build": "grunt build:js",
-    "test": "grunt build:js && babel-node ./node_modules/istanbul/lib/cli.js cover jasmine",
-    "check-coverage": "istanbul check-coverage --statements 86 --branches 77 --functions 78 --lines 86"
+    "lint": "tslint -c tslint.json 'src/**/*.{ts,tsx}'",
```
[the remark this answers is not in the document]
Comment (BramHeijmink) [reply package.json]: It is now the same.

[the remark this answers is not in the document]
Comment (BramHeijmink) [reply package.json]: Travis now runs ```yarn lint```.

**Comment by atimmer** [comment]
Jest 🎉

**Pull request merged**